modules/engine/src/model/model.ts

```diff
@@ -291,9 +289,33 @@ export class Model {
    * @note Triggers a pipeline rebuild / pipeline cache fetch on WebGPU
    */
   setBufferLayout(bufferLayout: BufferLayout[]): void {
-    if (bufferLayout !== this.bufferLayout) {
-      this.bufferLayout = bufferLayout;
-      this._setPipelineNeedsUpdate('bufferLayout');
+    this.bufferLayout = this._gpuGeometry
+      ? mergeBufferLayouts(bufferLayout, this._gpuGeometry.bufferLayout)
+      : bufferLayout;
+    this._setPipelineNeedsUpdate('bufferLayout');
+
+    // Recreate the pipeline
+    this.pipeline = this._updatePipeline();
+
+    const oldVertexArray = this.vertexArray;
+    // vertex array needs to be updated if we update buffer layout,
+    // but not if we update parameters
+    this.vertexArray = this.device.createVertexArray({
```

Comment:
Longer term, this pipeline recreation here may not be needed, as the vertex array doesn't actually need the pipeline, just the shader layout.

Currently on WebGL (but not on WebGPU) the shader layout is currently auto inferred from the linked program, but for WebGPU we always need to provide the shaderLayout as a prop to the model. (unless we write code to extract it from shader source, in which case we still don't need the pipeline).